File: mpd/Hooks/Hooks.swift
// mpd — Mpd.Hooks namespace
//
// Engine for the publish/subscribe hook system. Verb handlers fire
// typed Events; bash hook scripts at hooks/<event-name>.d/ inside
// container assets subscribe. See docs/HOOKS.md for vocabulary, the
// resource lifecycle model, the v1 event catalogue, and the
// acceptance test.
//
// Vocabulary:
//   Event    — Swift class declared per fire-point
//   Hook     — bash script that observes an event, lives on disk
//   Audience — list of container kinds an event reaches
//
// Adding an event: define `struct EventXxxYyy: Mpd.Hooks.Event` with
// the audience list, env vars, and failure mode. Call
// `try Mpd.Hooks.fire(EventXxxYyy(...))` from the verb handler. Drop
// hook scripts under the matching `hooks/<event-name>.d/` directory in
// the right asset layer.

import Foundation

extension Mpd {
    enum Hooks {}
}

// MARK: - Audience

extension Mpd.Hooks {
    /// Container kinds an event can reach. The dispatcher delivers an
    /// event to every running container of the given kind.
    enum Audience: Equatable {
        case runtime
        case database
        case service(String)
    }

    /// Failure semantics for an event class.
    enum FailureMode {
        /// Hook failure aborts the firing verb. Use for pre-conditions
        /// (e.g. "ensure DB is up before project start").
        case abort
        /// Hook failure logs but the verb proceeds. Use for cleanup-style
        /// and post-state events ("you can't fail to stop").
        case `continue`
    }
}

// MARK: - Event protocol

extension Mpd.Hooks {
    /// An Event is a typed Swift class declared per fire-point. Verbs
    /// construct an Event with relevant context and call `Mpd.Hooks.fire`.
    protocol Event {
        /// Event name in kebab-case. Auto-derived from the type name by
        /// stripping the `Event` prefix and converting to kebab-case.
        /// `EventProjectPreStart` → `project-pre-start`.
        static var name: String { get }

        /// Bumped when the env-var contract or audience list changes
        /// in a way that hooks should review. Diagnostics flag a bump
        /// at the next `mpd --setup`.
        static var revision: Int { get }

        /// Container kinds this event is delivered to.
        static var audiences: [Audience] { get }

        /// What happens to the firing verb when a hook script exits non-zero.
        static var onFailure: FailureMode { get }

        /// Per-script timeout. Default 30 s.
        static var timeout: TimeInterval { get }

        /// Event-specific context surfaced to hook scripts as
        /// `MPD_HOOK_<KEY>` env vars. Three standard vars
        /// (`MPD_HOOK_EVENT`, `MPD_HOOK_REVISION`, `MPD_HOOK_VERB`) are
        /// added by the dispatcher; this method provides only the typed
        /// extras specific to this event.
        var env: [String: String] { get }

        /// Containers reached by this event for a given audience. Default
        /// implementation enumerates all running containers of the audience
        /// kind (used by mpd-level events). Project-level events override
        /// to return just the project's runtime / DB.
        func containers(for audience: Audience) -> [String]
    }
}

// MARK: - Event default implementations

extension Mpd.Hooks.Event {
    static var name: String { Mpd.Hooks.derivedEventName(Self.self) }
    static var revision: Int { 1 }
    static var timeout: TimeInterval { 30 }

    func containers(for audience: Mpd.Hooks.Audience) -> [String] {
        Mpd.Hooks.runningContainers(for: audience)
    }
}

// MARK: - Name derivation

extension Mpd.Hooks {
    /// `EventProjectPreStart` → `project-pre-start`.
    static func derivedEventName<T>(_ type: T.Type) -> String {
        let typeName = String(describing: type)
        let stripped = typeName.hasPrefix("Event") ? String(typeName.dropFirst(5)) : typeName
        return kebabCase(stripped)
    }

    /// `MpdPreStop` → `mpd-pre-stop`.
    /// Inserts a hyphen before any uppercase letter that's preceded by
    /// a lowercase letter; lowercases the result.
    static func kebabCase(_ s: String) -> String {
        var out = ""
        let chars = Array(s)
        for i in 0..<chars.count {
            let ch = chars[i]
            if i > 0, ch.isUppercase, chars[i - 1].isLowercase {
                out.append("-")
            }
            out.append(ch)
        }
        return out.lowercased()
    }
}

// MARK: - Audience → running containers

extension Mpd.Hooks {
    /// Default audience → running container resolution. Used by the
    /// `Event` protocol's default `containers(for:)`. Project-level
    /// events override to scope to a single container.
    static func runningContainers(for audience: Audience) -> [String] {
        switch audience {
        case .runtime:
            return Mpd.Podman.ps(filter: "label=mpd.runtime")
                .filter { $0.State == "running" }
                .compactMap { $0.Names.first }
        case .database:
            return Mpd.Podman.ps(filter: "label=mpd.type=db")
                .filter { $0.State == "running" }
                .compactMap { $0.Names.first }
        case .service(let name):
            let cName = "mpd-service-\(name)"
            return Mpd.Podman.running(cName) ? [cName] : []
        }
    }
}

// MARK: - Fire dispatcher

extension Mpd.Hooks {
    /// Fire an event. Enumerates audience containers, invokes hook
    /// scripts in numeric-prefix order via `podman exec`, applies the
    /// event's failure mode.
    ///
    /// Throws when any `.abort`-mode hook fails. `.continue`-mode failures
    /// are logged but don't propagate.
    static func fire<E: Event>(_ event: E, verb: String = "") throws {
        let eventName = E.name
        var aborted: Error?

        for audience in E.audiences {
            let containers = event.containers(for: audience)
            for container in containers {
                do {
                    try fireOne(
                        event: event,
                        eventName: eventName,
                        audience: audience,
                        container: container,
                        verb: verb
                    )
                } catch {
                    if E.onFailure == .abort {
                        aborted = error
                        break
                    }
                    // .continue: fireOne already printed ✗ — keep going.
                }
            }
            if aborted != nil { break }
        }
        if let aborted { throw aborted }
    }

    /// Run all hook scripts for one event on one container.
    private static func fireOne<E: Event>(
        event: E,
        eventName: String,
        audience: Audience,
        container: String,
        verb: String
    ) throws {
        let scripts = (try? discoverScripts(audience: audience, container: container, eventName: eventName)) ?? []
        guard !scripts.isEmpty else { return }

        // Standard env vars + event-specific extras, all prefixed MPD_HOOK_.
        var podmanOptions: [String] = []
        var envVars: [String: String] = [
            "MPD_HOOK_EVENT": eventName,
            "MPD_HOOK_REVISION": String(E.revision),
            "MPD_HOOK_VERB": verb,
        ]
        for (k, v) in event.env {
            envVars["MPD_HOOK_\(k)"] = v
        }
        for (k, v) in envVars {
            podmanOptions.append("--env")
            podmanOptions.append("\(k)=\(v)")
        }

        for script in scripts {
            let label = "[\(container)] \(eventName)/\(script.basename)"
            let started = Date()
            // Enforced, not just declared: a hook that never returns
            // would otherwise hang the verb that fired it — and for
            // mpd-pre-stop, hang VM shutdown.
            let rc = Mpd.Podman.execWithTimeout(container, options: podmanOptions,
                                                ["bash", script.containerPath],
                                                timeout: E.timeout)
            let elapsed = Int(Date().timeIntervalSince(started))
            if rc == 0 {
                print("  \(label) ✓ (\(elapsed)s)")
            } else if rc == Mpd.VM.exitTimedOut {
                print("  \(label) ✗ timed out after \(Int(E.timeout))s")
                throw RuntimeError(
                    "Hook \(eventName) timed out on \(container) after \(Int(E.timeout))s")
            } else {
                print("  \(label) ✗ exit \(rc) (\(elapsed)s)")
                throw RuntimeError("Hook \(eventName) failed on \(container) (exit \(rc))")
            }
        }
    }
}

// MARK: - Script discovery

extension Mpd.Hooks {
    fileprivate struct DiscoveredScript {
        /// Filename only — e.g. `10-graceful-stop`.
        let basename: String
        /// Path inside the container — e.g.
        /// `/opt/mpd/assets/databases/postgres/hooks/mpd-pre-stop.d/10-graceful-stop`.
        let containerPath: String
    }

    /// Walk asset directories for hook scripts matching this event +
    /// container. Layered: base → runtime → type for runtimes; per-engine
    /// for databases. Returns scripts in execution order (layer first,
    /// then alphabetical within a directory — numeric prefixes order
    /// scripts inside a single layer).
    fileprivate static func discoverScripts(
        audience: Audience,
        container: String,
        eventName: String
    ) throws -> [DiscoveredScript] {
        let assets = try Mpd.VM.assetsPath()
        var dirs: [(hostDir: String, containerDir: String)] = []

        switch audience {
        case .runtime:
            // Runtime containers have `<assets>:/opt/mpd/assets:ro` bind-mounted.
            // Layered: base + per-runtime. Type-level layer (per-project)
            // is project-event territory and not yet wired in v1.
            dirs.append((
                "\(assets)/runtime-base/hooks/\(eventName).d",
                "/opt/mpd/assets/runtime-base/hooks/\(eventName).d"
            ))
            let runtimeName = Mpd.Podman.label(container, "mpd.name")
            if !runtimeName.isEmpty {
                dirs.append((
                    "\(assets)/runtimes/\(runtimeName)/hooks/\(eventName).d",
                    "/opt/mpd/assets/runtimes/\(runtimeName)/hooks/\(eventName).d"
                ))
            }
        case .database:
            // DB containers get the same `<assets>:/opt/mpd/assets:ro` bind
            // mount — added in `Mpd.Runtime.DB.ensure`. Per-engine only;
            // there's no `database-base` layer (DB images are stock).
            let engine = Mpd.Podman.label(container, "mpd.db.engine")
            if !engine.isEmpty {
                dirs.append((
                    "\(assets)/databases/\(engine)/hooks/\(eventName).d",
                    "/opt/mpd/assets/databases/\(engine)/hooks/\(eventName).d"
                ))
            }
        case .service(let name):
            dirs.append((
                "\(assets)/services/\(name)/hooks/\(eventName).d",
                "/opt/mpd/assets/services/\(name)/hooks/\(eventName).d"
            ))
        }

        var scripts: [DiscoveredScript] = []
        let fm = FileManager.default
        for (hostDir, containerDir) in dirs {
            guard fm.fileExists(atPath: hostDir) else { continue }
            let names = (try? fm.contentsOfDirectory(atPath: hostDir)) ?? []
            for n in names.sorted() {
                // Only `*.sh`. Without an extension filter every file in
                // the directory is executed — an editor backup
                // (`10-foo.sh~`), a `.bak`, a stray `.swp`, or a README
                // would all be run through bash. Requiring the extension
                // makes "is this a hook?" answerable by looking, and a
                // hook that needs a compiled helper just execs it from a
                // one-line wrapper.
                guard n.hasSuffix(".sh") else { continue }
                let host = "\(hostDir)/\(n)"
                var isDir: ObjCBool = false
                guard fm.fileExists(atPath: host, isDirectory: &isDir),
                      !isDir.boolValue else { continue }
                scripts.append(DiscoveredScript(
                    basename: n,
                    containerPath: "\(containerDir)/\(n)"
                ))
            }
        }
        return scripts
    }
}

File: go/internal/hooks/hooks.go
// Package hooks dispatches mpd's typed lifecycle events to bash scripts
// in the asset tree.
//
// Events, hooks and audiences are defined in docs/hooks.md. The event
// names, the MPD_HOOK_* environment and the ordering are a public
// contract for scripts outside this repo and must not drift.
package hooks

import (
	"context"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/mutms/mpd/go/internal/exec"
	"github.com/mutms/mpd/go/internal/podman"
)

// AssetsDir is the bind-mounted asset tree, identical on the VM and
// inside every container — so a host-side scan yields a path the
// container can execute.
const AssetsDir = "/opt/mpd/assets"

// assetsDir is what discovery reads; tests point it at a fixture tree.
var assetsDir = AssetsDir

// DefaultTimeout bounds a hook script that declares none.
//
// Caveat, verified: killing `podman exec` does NOT kill the process
// inside the container. A timed-out hook stops blocking mpd, but its
// script keeps running until it finishes or the container stops.
const DefaultTimeout = 30 * time.Second

// StopTimeout bounds mpd-pre-stop, where a database flushing pending IO
// legitimately takes longer than a normal hook.
const StopTimeout = 120 * time.Second

// SetupTimeout bounds mpd-post-setup, whose hooks install things.
const SetupTimeout = 10 * time.Minute

// AudienceKind selects which containers an event fires into.
type AudienceKind int

const (
	// AudienceRuntime: the project's runtime container.
	AudienceRuntime AudienceKind = iota
	// AudienceDatabase: the project's database container.
	AudienceDatabase
	// AudienceService: a named always-on service container.
	AudienceService
	// AudienceVM: the VM host itself, for work no container can do.
	AudienceVM
)

// VMTarget labels AudienceVM output where others print a container name.
const VMTarget = "vm"

// FailureMode decides what a failing hook does to the firing verb.
type FailureMode int

const (
	// Abort: hook failure fails the verb. For preconditions.
	Abort FailureMode = iota
	// Continue: hook failure is logged and the verb proceeds. For
	// cleanup and post-state events: you cannot fail to stop.
	Continue
)

// Event is one typed lifecycle moment.
type Event struct {
	// Name is the kebab-case event name, matching the `<name>.d` directory.
	Name string
	// Revision is surfaced as MPD_HOOK_REVISION so a hook can tell which
	// contract it is being called under.
	Revision int
	// Timeout bounds each script. Zero means DefaultTimeout.
	Timeout time.Duration
	// Audiences the event fires into, in order.
	Audiences []AudienceKind
	// OnFailure decides whether a failing hook aborts the verb.
	OnFailure FailureMode
	// Env holds event-specific values, exported as MPD_HOOK_<KEY>.
	Env map[string]string
	// Containers resolves an audience to the containers to fire into.
	// Returning none is normal: a project with no database has nothing
	// to fire a database-audience event into.
	Containers func(AudienceKind) []string
	// ServiceName scopes AudienceService to one service's assets.
	ServiceName string
	// User is who runtime hooks run as; empty means root. AudienceRuntime
	// only — DB and service containers have no dev user.
	User string
}

// Fire runs every hook for an event, honouring its failure mode.
//
// Ordering is part of the contract: audiences in declared order,
// containers within an audience, then scripts alphabetically — which is
// what makes numeric prefixes (10-, 90-) meaningful.
func Fire(ctx context.Context, out io.Writer, ev Event, verb string, p *podman.Client) error {
	var aborted error

	for _, audience := range ev.Audiences {
		for _, container := range ev.targets(audience) {
			err := fireOne(ctx, out, ev, audience, container, verb, p)
			if err == nil {
				continue
			}
			if ev.OnFailure == Abort {
				aborted = err
				break
			}
			// Continue mode: fireOne already reported the failure.
		}
		if aborted != nil {
			break
		}
	}
	return aborted
}

// targets resolves an audience. AudienceVM is always the one VM.
func (ev Event) targets(a AudienceKind) []string {
	if a == AudienceVM {
		return []string{VMTarget}
	}
	return ev.Containers(a)
}

func fireOne(ctx context.Context, out io.Writer, ev Event, audience AudienceKind,
	container, verb string, p *podman.Client) error {

	scripts := discover(ctx, p, ev, audience, container)
	if len(scripts) == 0 {
		return nil
	}

	env := hookEnv(ev, verb)
	opts := execOptions(ev, audience, env)
	envList := hookEnvList(ev, verb)

	timeout := ev.Timeout
	if timeout == 0 {
		timeout = DefaultTimeout
	}

	for _, script := range scripts {
		label := fmt.Sprintf("[%s] %s/%s", container, ev.Name, script.Basename)
		started := time.Now()

		// The timeout is enforced: a hook that never returns would hang
		// the verb, and for mpd-pre-stop, hang VM shutdown.
		runCtx, cancel := context.WithTimeout(ctx, timeout)
		var code int
		var err error
		if audience == AudienceVM {
			code, err = exec.Run(runCtx, exec.Cmd{
				Name: "bash", Args: []string{script.Path}, Env: envList,
			})
		} else {
			code, err = p.ExecWithOptions(runCtx, container, opts, "bash", script.Path)
		}
		timedOut := runCtx.Err() == context.DeadlineExceeded
		cancel()

		elapsed := int(time.Since(started).Seconds())
		switch {
		case timedOut:
			fmt.Fprintf(out, "  %s ✗ timed out after %ds\n", label, int(timeout.Seconds()))
			return fmt.Errorf("Hook %s timed out on %s after %ds",
				ev.Name, container, int(timeout.Seconds()))
		case err == nil && code == 0:
			fmt.Fprintf(out, "  %s ✓ (%ds)\n", label, elapsed)
		default:
			fmt.Fprintf(out, "  %s ✗ exit %d (%ds)\n", label, code, elapsed)
			return fmt.Errorf("Hook %s failed on %s (exit %d)", ev.Name, container, code)
		}
	}
	return nil
}

// hookEnv is the MPD_HOOK_* environment: the standard three plus the
// event's own values, prefixed.
func hookEnv(ev Event, verb string) map[string]string {
	env := map[string]string{
		"MPD_HOOK_EVENT":    ev.Name,
		"MPD_HOOK_REVISION": fmt.Sprint(ev.Revision),
		"MPD_HOOK_VERB":     verb,
	}
	for k, v := range ev.Env {
		env["MPD_HOOK_"+k] = v
	}
	return env
}

// hookEnvList renders that as KEY=VALUE, for the VM audience.
func hookEnvList(ev Event, verb string) []string {
	env := hookEnv(ev, verb)
	var list []string
	for _, k := range sortedKeys(env) {
		list = append(list, k+"="+env[k])
	}
	return list
}

// execOptions renders the podman exec flags. Sorted, so argv is
// deterministic.
func execOptions(ev Event, audience AudienceKind, env map[string]string) []string {
	if env == nil {
		env = hookEnv(ev, "")
	}
	var opts []string
	for _, k := range sortedKeys(env) {
		opts = append(opts, "--env", k+"="+env[k])
	}
	// Without --user a runtime hook lands in /root, not the dev's home.
	if audience == AudienceRuntime && ev.User != "" {
		opts = append(opts, "--user", ev.User)
	}
	return opts
}

func sortedKeys(m map[string]string) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// Script is one discovered hook.
type Script struct {
	Basename string
	// Path is the same string inside a container and on the VM, since
	// /opt/mpd is mounted at the same path.
	Path string
}

// discover finds the hook scripts for an event on one audience.
//
// Which assets apply comes from the CONTAINER's labels, not the event:
// mpd-pre-stop fires on every running database at once, and a single
// engine on the event would send one engine's hooks into another. Only
// `*.sh` is considered, so editor backups and READMEs are never handed
// to bash (see docs/hooks.md).
func discover(ctx context.Context, p *podman.Client, ev Event, audience AudienceKind, container string) []Script {
	var dirs []string

	switch audience {
	case AudienceRuntime:
		// Deliberately no project-type layer here: firing one would
		// break the v1 hook contract (docs/hooks.md).
		dirs = append(dirs,
			filepath.Join(assetsDir, "runtime", "hooks", ev.Name+".d"))
	case AudienceDatabase:
		// Per-engine only; DB images are stock, so there is no base layer.
		if engine := containerLabel(ctx, p, container, "mpd.db.engine"); engine != "" {
			dirs = append(dirs, filepath.Join(assetsDir, "databases", engine, "hooks", ev.Name+".d"))
		}
	case AudienceService:
		if ev.ServiceName != "" {
			dirs = append(dirs, filepath.Join(assetsDir, "services", ev.ServiceName, "hooks", ev.Name+".d"))
		}
	case AudienceVM:
		dirs = append(dirs,
			filepath.Join(assetsDir, "vm", "hooks", ev.Name+".d"))
	}

	var scripts []Script
	for _, dir := range dirs {
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}
		var names []string
		for _, e := range entries {
			if e.IsDir() || !strings.HasSuffix(e.Name(), ".sh") {
				continue
			}
			names = append(names, e.Name())
		}
		sort.Strings(names)
		for _, n := range names {
			scripts = append(scripts, Script{
				Basename: n,
				Path:     filepath.Join(dir, n),
			})
		}
	}
	return scripts
}

// containerLabel reads a label, tolerating a nil client so discovery
// can be tested without podman.
func containerLabel(ctx context.Context, p *podman.Client, container, key string) string {
	if p == nil || container == "" {
		return ""
	}
	return p.Label(ctx, container, key)
}
